sky/provision/kubernetes/network_utils.py

```diff
@@ -0,0 +1,206 @@
+        ingress_name,
+        namespace,
+        ingress_spec,
+        _request_timeout=kubernetes.API_TIMEOUT)
+
+
+def delete_namespaced_ingress(namespace: str, ingress_name: str) -> None:
+    """Deletes an ingress resource."""
+    networking_api = kubernetes.networking_api()
+    try:
+        networking_api.delete_namespaced_ingress(
+            ingress_name, namespace, _request_timeout=kubernetes.API_TIMEOUT)
+    except kubernetes.get_kubernetes().client.ApiException as e:
+        if e.status == 404:
+            raise exceptions.PortDoesNotExistError(
+                f'Port {ingress_name.split("--")[-1]} does not exist.')
+        raise e
+
+
+def create_or_replace_namespaced_service(
+        namespace: str, service_name: str,
+        service_spec: Dict[str, Union[str, int]]) -> None:
+    """Creates a service resource for the specified service."""
+    core_api = kubernetes.core_api()
+
+    try:
+        core_api.read_namespaced_service(
+            service_name, namespace, _request_timeout=kubernetes.API_TIMEOUT)
+    except kubernetes.get_kubernetes().client.ApiException as e:
+        if e.status == 404:
+            core_api.create_namespaced_service(
+                namespace,
+                service_spec,
+                _request_timeout=kubernetes.API_TIMEOUT)
+            return
+        raise e
+
+    core_api.replace_namespaced_service(service_name,
+                                        namespace,
+                                        service_spec,
+                                        _request_timeout=kubernetes.API_TIMEOUT)
+
+
+def delete_namespaced_service(namespace: str, service_name: str) -> None:
+    """Deletes a service resource."""
+    core_api = kubernetes.core_api()
+
+    try:
+        core_api.delete_namespaced_service(
+            service_name, namespace, _request_timeout=kubernetes.API_TIMEOUT)
+    except kubernetes.get_kubernetes().client.ApiException as e:
+        if e.status == 404:
+            raise exceptions.PortDoesNotExistError(
+                f'Port {service_name.split("--")[-1]} does not exist.')
+        raise e
+
+
+def ingress_controller_exists(ingress_class_name: str = 'nginx') -> bool:
+    """Checks if an ingress controller exists in the cluster."""
+    networking_api = kubernetes.networking_api()
+    ingress_classes = networking_api.list_ingress_class(
+        _request_timeout=kubernetes.API_TIMEOUT).items
+    return any(
+        map(lambda item: item.metadata.name == ingress_class_name,
+            ingress_classes))
+
+
+def get_external_ip_and_ports(
+        namespace: str) -> Tuple[Optional[str], Optional[Tuple[int, int]]]:
+    """Returns external ip and ports for the ingress controller."""
+    core_api = kubernetes.core_api()
+    ingress_services = [
+        item for item in core_api.list_namespaced_service(
+            namespace, _request_timeout=kubernetes.API_TIMEOUT).items
+        if item.spec.type == 'LoadBalancer'
+    ]
```

Comment:
I was trying this on a [kind cluster with nginx ingress](https://kind.sigs.k8s.io/docs/user/ingress/#ingress-nginx) and querying the endpoint was failing with:

```
(base) ➜  ~ sky status test --endpoint 8888
ValueError: Port 8888 not exposed yet.
If the cluster was recently started, please retry after a while.
```

Instead of finding a service of type `LoadBalancer` and then setting `ingress_service = ingress_services[0]` at L182, can we instead directly filter for service with `item.metadata.name == ingress-nginx-controller`? Might also need to update some of the downstream code here..

Reply:
Repro:

1. Update your kind config generate (generate_kind_config.py) with this gist ([follows the kind guide for setting up nginx](https://kind.sigs.k8s.io/docs/user/ingress/#ingress-nginx)): https://gist.github.com/romilbhardwaj/128dc1bb35ccf454962224137008502e
2. Run `sky local up`
3. Install nginx ingress for kind - `kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/kind/deploy.yaml`
4.  Launch cluster`sky launch -c test test.yaml`
```
# test.yaml
resources:
  ports:
    - 8888

setup: |
  echo hi

run: |
  python -m http.server 8888
```
5. Run `sky status test --endpoint 8888`

Note that I can manually enter `http://localhost/skypilot/test-2ea4/8888` in my browser and it works, just the endpoint fetching logic needs to be fixed for on-prem/non-loadbalancer ingresses.